torch/_logging/_internal.py

```diff
@@ -0,0 +1,424 @@
+    if settings == "":
+        return dict()
+
+    if not _validate_settings(settings):
+        raise ValueError(
+            f"Invalid log settings: {settings}, must be a comma separated list of registerered log or artifact names."
+        )
+
+    settings = re.sub(r"\s+", "", settings)
+    log_names = settings.split(",")
+
+    def get_name_level_pair(name):
+        clean_name = name.replace(INCR_VERBOSITY_CHAR, "")
+        clean_name = clean_name.replace(DECR_VERBOSITY_CHAR, "")
+
+        if name[0] == INCR_VERBOSITY_CHAR:
+            level = logging.DEBUG
+        elif name[0] == DECR_VERBOSITY_CHAR:
+            level = logging.ERROR
+        else:
+            level = logging.INFO
+
+        return clean_name, level
+
+    log_state = LogState()
+    for name in log_names:
+        name, level = get_name_level_pair(name)
+        if log_registry.is_log(name):
+            assert level is not None
+            log_qname = log_registry.log_alias_to_log_qname[name]
+            log_state.enable_log(log_qname, level)
+        elif log_registry.is_artifact(name):
+            log_state.enable_artifact(name)
+        elif _is_valid_module(name):
+            if not _has_registered_parent(name):
+                log_registry.register_log(name, name)
+            else:
+                log_registry.register_child_log(name)
+            log_state.enable_log(name, level)
+        else:
+            raise ValueError(
+                f"Invalid log settings: '{settings}', must be a comma separated list of log or artifact names."
+            )
+
+    return log_state
+
+
+def _is_valid_module(qname):
+    try:
+        __import__(qname)
+        return True
+    except ImportError:
+        return False
+
+
+def _update_log_state_from_env():
+    global log_state
+    log_setting = os.environ.get(LOG_ENV_VAR, None)
+    if log_setting is not None:
+        log_state = _parse_log_settings(log_setting)
+
+
+def _has_registered_parent(log_qname):
+    cur_log = logging.getLogger(log_qname)
+
+    registered_log_qnames = log_registry.get_log_qnames()
+
+    while cur_log.parent:
+        if cur_log.name in registered_log_qnames:
+            return True
+        cur_log = cur_log.parent
+
+    return False
+
+
+# setup custom handlers
+# if the log level of a component is set to INFO, setup
+# an additional handler to print those messages, because
+# the debug handler is what handles custom objects like guards,
+# bytecode, etc.
+# if the log level of a component is set to DEBUG, allow all
+# string messages and allowed types (other than those off by default)
+def _setup_handlers(create_handler_fn, log):
+    debug_handler = _tag_handler(create_handler_fn())
+    debug_handler.setFormatter(DEFAULT_FORMATTER)
+    debug_handler.setLevel(logging.DEBUG)
+    log.addHandler(debug_handler)
+
+
+# mark handlers that we've created
+# so we don't modify user handlers
+def _tag_handler(handler):
+    handler.__torch_log_handler = True
```

Comment:
NB: using double underscore here is confusing, because it implies there is mangling, but in fact there is no mangling (as you can see from the hasattr call).

FWIW, since handlers are singletons, you could also just chuck them in weak set and identify them that way too. Or just subclass the handler.

Reply:
Yeah I'll just do a weakset

Reply:
Did the weakset